# Load Libraries

In [0]:
import pandas as pd
import numpy as np
import plotly.express as px
import plotly.graph_objects as go
from sklearn.impute import SimpleImputer


# Load Data

*Annotation: The command 'pd.read' when importing the raw data file from my GitHub repository resulted in an error message, so I uploaded the raw data file on my company's webserver:*

In [55]:
!wget https://www.attax.at/bank-additional-full.csv

--2020-06-01 20:06:22--  https://www.attax.at/bank-additional-full.csv
Resolving www.attax.at (www.attax.at)... 213.145.224.160
Connecting to www.attax.at (www.attax.at)|213.145.224.160|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 5834924 (5.6M) [text/csv]
Saving to: ‘bank-additional-full.csv.1’

bank-additional-ful 100%[===================>]   5.56M  4.01MB/s    in 1.4s    

2020-06-01 20:06:25 (4.01 MB/s) - ‘bank-additional-full.csv.1’ saved [5834924/5834924]



In [0]:
df = pd.read_csv("bank-additional-full.csv", sep=";")

In [57]:
df.head()

age        job  marital  ... euribor3m nr.employed   y
0   56  housemaid  married  ...     4.857      5191.0  no
1   57   services  married  ...     4.857      5191.0  no
2   37   services  married  ...     4.857      5191.0  no
3   40     admin.  married  ...     4.857      5191.0  no
4   56   services  married  ...     4.857      5191.0  no

[5 rows x 21 columns]

## Rename Columns for Convenience

In [0]:
df.columns = [x.replace('.', '_') for x in df.columns]

In [59]:
df.head()

age        job  marital  ... euribor3m nr_employed   y
0   56  housemaid  married  ...     4.857      5191.0  no
1   57   services  married  ...     4.857      5191.0  no
2   37   services  married  ...     4.857      5191.0  no
3   40     admin.  married  ...     4.857      5191.0  no
4   56   services  married  ...     4.857      5191.0  no

[5 rows x 21 columns]

In [60]:
df.tail()

age          job  marital  ... euribor3m nr_employed    y
41183   73      retired  married  ...     1.028      4963.6  yes
41184   46  blue-collar  married  ...     1.028      4963.6   no
41185   56      retired  married  ...     1.028      4963.6   no
41186   44   technician  married  ...     1.028      4963.6  yes
41187   74      retired  married  ...     1.028      4963.6   no

[5 rows x 21 columns]

In [61]:
df.columns.to_list()

['age',
 'job',
 'marital',
 'education',
 'default',
 'housing',
 'loan',
 'contact',
 'month',
 'day_of_week',
 'duration',
 'campaign',
 'pdays',
 'previous',
 'poutcome',
 'emp_var_rate',
 'cons_price_idx',
 'cons_conf_idx',
 'euribor3m',
 'nr_employed',
 'y']

In [62]:
df.describe()

age      duration  ...     euribor3m   nr_employed
count  41188.00000  41188.000000  ...  41188.000000  41188.000000
mean      40.02406    258.285010  ...      3.621291   5167.035911
std       10.42125    259.279249  ...      1.734447     72.251528
min       17.00000      0.000000  ...      0.634000   4963.600000
25%       32.00000    102.000000  ...      1.344000   5099.100000
50%       38.00000    180.000000  ...      4.857000   5191.000000
75%       47.00000    319.000000  ...      4.961000   5228.100000
max       98.00000   4918.000000  ...      5.045000   5228.100000

[8 rows x 10 columns]

## Check Missing Values

In [63]:
df.isna().sum()

age               0
job               0
marital           0
education         0
default           0
housing           0
loan              0
contact           0
month             0
day_of_week       0
duration          0
campaign          0
pdays             0
previous          0
poutcome          0
emp_var_rate      0
cons_price_idx    0
cons_conf_idx     0
euribor3m         0
nr_employed       0
y                 0
dtype: int64

Apparently, there are no missing values. However, I will check for "hidden" missing values later.

## Dropping duplicates

In [64]:
df.shape

(41188, 21)

In [0]:
df = df.drop_duplicates()

In [66]:
df.shape

(41176, 21)

In [67]:
df.head()

age        job  marital  ... euribor3m nr_employed   y
0   56  housemaid  married  ...     4.857      5191.0  no
1   57   services  married  ...     4.857      5191.0  no
2   37   services  married  ...     4.857      5191.0  no
3   40     admin.  married  ...     4.857      5191.0  no
4   56   services  married  ...     4.857      5191.0  no

[5 rows x 21 columns]

There were 12 duplicates.

# Getting an overview of the features,  to determine which might be not relevant at all

In [68]:
df.job.value_counts()

admin.           10419
blue-collar       9253
technician        6739
services          3967
management        2924
retired           1718
entrepreneur      1456
self-employed     1421
housemaid         1060
unemployed        1014
student            875
unknown            330
Name: job, dtype: int64

In [69]:
df.pdays.value_counts(ascending=True)

27         1
25         1
20         1
26         1
21         2
22         3
19         3
18         7
17         8
16        11
0         15
8         18
14        20
15        24
1         26
11        28
13        36
5         46
10        52
12        58
7         60
2         61
9         64
4        118
6        412
3        439
999    39661
Name: pdays, dtype: int64

In [70]:
df.pdays.describe()

count    41176.000000
mean       962.464810
std        186.937102
min          0.000000
25%        999.000000
50%        999.000000
75%        999.000000
max        999.000000
Name: pdays, dtype: float64

In [71]:
df.marital.value_counts()

married     24921
single      11564
divorced     4611
unknown        80
Name: marital, dtype: int64

In [72]:
df.education.value_counts()

university.degree      12164
high.school             9512
basic.9y                6045
professional.course     5240
basic.4y                4176
basic.6y                2291
unknown                 1730
illiterate                18
Name: education, dtype: int64

In [73]:
df["default"].value_counts()


no         32577
unknown     8596
yes            3
Name: default, dtype: int64

In [74]:
df.housing.value_counts()

yes        21571
no         18615
unknown      990
Name: housing, dtype: int64

In [75]:
df.loan.value_counts()

no         33938
yes         6248
unknown      990
Name: loan, dtype: int64

In [76]:
df.contact.value_counts()

cellular     26135
telephone    15041
Name: contact, dtype: int64

In [77]:
df.poutcome.value_counts()

nonexistent    35551
failure         4252
success         1373
Name: poutcome, dtype: int64

In [78]:
df.y.value_counts()

no     36537
yes     4639
Name: y, dtype: int64

In [79]:
df.campaign.value_counts()

1     17634
2     10568
3      5340
4      2650
5      1599
6       979
7       629
8       400
9       283
10      225
11      177
12      125
13       92
14       69
17       58
15       51
16       51
18       33
20       30
19       26
21       24
22       17
23       16
24       15
27       11
29       10
25        8
26        8
28        8
30        7
31        7
35        5
33        4
32        4
34        3
40        2
42        2
43        2
37        1
39        1
41        1
56        1
Name: campaign, dtype: int64

In [80]:
df.previous.value_counts()

0    35551
1     4561
2      754
3      216
4       70
5       18
6        5
7        1
Name: previous, dtype: int64

In [81]:
df.day_of_week.value_counts()

thu    8618
mon    8512
wed    8134
tue    8086
fri    7826
Name: day_of_week, dtype: int64

In [82]:
df.month.value_counts()

may    13767
jul     7169
aug     6176
jun     5318
nov     4100
apr     2631
oct      717
sep      570
mar      546
dec      182
Name: month, dtype: int64

In [83]:
df.month.nunique()

10

**Findings: In some categorical columns, there are missing values, referred to as "unknown". Feature 'pdays' (extremely) most common value is 999, however, the other values vary from 3 to 27. So it will be dropped. Feature 'poutcome' has an extremely high number of 'nonexistent'. It stands for "predicted outcome", but since we know about the real outcome in feature 'y', it will be dropped.**

#Impute missing values in categorical columns

In [84]:
df.columns

Index(['age', 'job', 'marital', 'education', 'default', 'housing', 'loan',
       'contact', 'month', 'day_of_week', 'duration', 'campaign', 'pdays',
       'previous', 'poutcome', 'emp_var_rate', 'cons_price_idx',
       'cons_conf_idx', 'euribor3m', 'nr_employed', 'y'],
      dtype='object')

In [0]:
categorical_columns_missing = ["job", "marital", "education", "housing", "loan"]


In [86]:
(df[categorical_columns_missing] =="unknown").sum()

job           330
marital        80
education    1730
housing       990
loan          990
dtype: int64

In [0]:
impute_missing = SimpleImputer(missing_values="unknown", strategy="most_frequent")

df[categorical_columns_missing] = impute_missing.fit_transform(df[categorical_columns_missing])

In [88]:
(df[categorical_columns_missing] =="unknown").sum()


job          0
marital      0
education    0
housing      0
loan         0
dtype: int64

# Dropping features 'poutcome' and 'pdays'

In [89]:
df.shape

(41176, 21)

In [0]:
df = df.drop(labels='poutcome', axis=1)

In [0]:
df = df.drop(labels='pdays', axis=1)

In [93]:
df.shape

(41176, 19)

# Exporting cleaned dataframe as csv-file

*Annotation: I exported the clean data file to my Google drive account, as I did not manage it to export it to my GitHub repository.*

In [0]:
from google.colab import drive

In [95]:
drive.mount('/drive')

Drive already mounted at /drive; to attempt to forcibly remount, call drive.mount("/drive", force_remount=True).


In [0]:
df.to_csv('/drive/My Drive/Colab Notebooks/clean.csv', index=False)